omnikinverter/omnikinverter.py

```diff
@@ -142,6 +148,8 @@ async def close(self) -> None:
         """Close open client session."""
         if self.session and self._close_session:
             await self.session.close()
+            self.session = None
+            self._close_session = False
```

Comment:
Is this really necessary? `_close_session` is by default already **False** with every new request.

Reply:
It looks like it is a class variable. so it is only set to `False` when instantiating a new `OmnikConverter` I think? 
The class is kept for all requests being done so `_close_session` stays around.

Since it was set to `True` because a manual session was constructed, I am restoring it to `False` when getting rid of the manual session to keep it symmetrical.

Reply:
Wouldn't it be better to just put this in the finally if statement?

Reply:
That should work also.
I tried moving it just now, but coverage is now complaining about the `session.close()` call in the `close()` function and I cannot figure out why :(

Reply:
Still wonder if all this is necessary 🤔 If you don't pass a session and R66 sets `_close_session` to **True**, also means that it will pass the if statement in the close function and call `await self.session.close()`

Reply:
The close session is by default called in the __aexit__ magic method and so far i know this is on every request.

Reply:
But `aexit` (with `aenter`) is only called when using a context manager (so within an `async with` block), but the HA integration is not using a contextmanager, which means `aexit` never gets called.

Reply:
But the package is using [asyn with](https://github.com/klaasnicolaas/python-omnikinverter/pull/95/files#diff-521682543174b52631cd8f53b94cd68e51f00db32ac413a2a04651e41afc94ebR83), the HA integration has nothing to do with that.

Reply:
I only see it used in this line:
```
async with async_timeout.timeout(self.request_timeout):
```
But doesn't that only apply to the timeout?
Don't you need something like `async with <instance of OmnikInverter>` to get the `aenter` and `aexit` of the OmnikInverter class called ?

Reply:
It applies to everything inside the `async with` block

https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-3.html
"If the code-block of the async with statement reaches its end without an exception then __aexit__ will be called"

Reply:
Weird, I am pretty sure I got errors about connections not being closed while I was testing wtihout the additional close logic. Will need to retest again. 
Unfortunately it is dark when I get home so that will be in the weekend.

Reply:
I finally got some time to do some experiments this afternoon. But it seems the `__aexit__` method is only called on the class that is used in the `with Class as c` statement.

I went back to the base integration and put some logging statements in the methods of the `OmnikInverter` class. Then I used the `test_output.py` script (slightly modified to take commandline params and add an additional experiment)


The standard code like below

```python
async with OmnikInverter(<params>) as client:
    inverter: OmnikInverter = await client.inverter()
    device: OmnikInverter = await client.device()
    <print statements>
```
resulted in output like this:

```log
DEBUG:root:> __aenter__
DEBUG:root:> inverter
DEBUG:root:> request
DEBUG:root:Creating internal ClientSession
DEBUG:root:Session in use is: <aiohttp.client.ClientSession object at 0x7f673d4422e0>
DEBUG:root:> device
DEBUG:root:> request
DEBUG:root:Session in use is: <aiohttp.client.ClientSession object at 0x7f673d4422e0>
DEBUG:root:> __aexit__
DEBUG:root:> close
DEBUG:root:Closing internal ClientSession
<tracebacks because stick does not like requests made in same session removed>
```

The same session is used for both requests. The  `__aenter__` and `__aexit__` are only called once each when entering/exiting the  `async with OmnikInverter(<params>) as client:` block.


As an alternative I tried manually creating the class (instead of using the `with` block) and then call the `inverter`, `device` and `close` methods. This resulted in these calls.

```
DEBUG:root:> inverter
DEBUG:root:> request
DEBUG:root:Creating internal ClientSession
DEBUG:root:Session in use is: <aiohttp.client.ClientSession object at 0x7fe5312ebfa0>
DEBUG:root:> device
DEBUG:root:> request
DEBUG:root:Session in use is: <aiohttp.client.ClientSession object at 0x7fe5312ebfa0>
<tracebacks because stick does not like requests made in same session removed>
```

Here the __aenter__ and __aexit__ methods are not called at all.

For reference code used to get the logs above can be found on this branch https://github.com/mvdwetering/python-omnikinverter/tree/experiments 
Run like `./test_output.py 192.168.178.32 html username password --loglevel DEBUG ` (sourcetype username password are optional)


So it seems really needed to manually close/cleanup the sessions after eacht request to avoid reuse. I don't see what I could have done wrong in these experiments

Reply:
Seems clear to me now regarding the magic methods, the code as it is now should indeed be fine. I still have to test with other inverters, but then we'll just merge this PR soon.

Thank you so much for sorting it out, it clarified a lot!